### PR TITLE
ENH: Species support in Database and TDB read/write

### PR DESCRIPTION
**Pull request by bocklund**
Species in a TDB are parsed in the `Specie` class and added to the `species` set attribute of the Database.

The Specie class has several attributes:

- `name`: the name of the specie
- `constituents`: a dictionary of `{element: composition}` of the specie
- `charge`: integer charge. Defaults to zero if none specified.

As discussed in the Thermo-Calc database manager guide, elements entered via the ELEMENT command are automatically entered as species. We match that behavior here. The constituents of an element are just that element with a fraction of 1 and the charge is 0.

In principle, the constituents dictionary will be used in the solver to link the specie to how it contributes to the overall mass balance conditions.

Tests are included that test the basic functionality. I think the examples I chose were representative, and I'm unaware of any edge cases that I missed.

This could be merged immediately to better support parsing of species, but no changes were made to the solver and the species defined here should not affect calculations in any way.

@richardotis comments? suggestions?

### CODE REVIEW

**Comment by bocklund** [comment]
This would move towards closing #123 (the Unary 50 TDB parses, but fails in writing, in what seems to be an issue unrelated to species) and check off the first two tasks in #110

**Comment by richardotis** [comment]
What about using SPECIES in PARAMETERs? Does that parse correctly?

**Comment by bocklund** [comment]
IIRC,it already does, and we don't enforce the need for the values to be elements. But I can write a test to verify

**Comment by bocklund** [comment]
Parsed correctly, but did not write correctly. Fixed with test in newest changes.

**Comment by richardotis** [comment]
Since this is backwards-compatible with the existing solver implementation I'm okay with not solving this in this issue, but I'd like to do some thinking about how species should be supported in `Model`, `calculate` and ultimately `equilibrium`.

**Comment by bocklund** [comment]
I agree. I think the species class might also be the correct place to keep our assumptions about the species, e.g. vacancies, so we can get away from implicit assumptions throughout the codebase.

For the most part, I envision that everything in `Model`, `calculate` and `equilibrium` will be based on species instead of elements. The only part that will depend on elements is the mass balance. 

I think most of the changes are relatively straightforward replacement of element with species in `Model` and `calculate`, but I imagine we'll have to make more complicated changes for supporting stepping in species space rather than element space in `equilibrium` and our visualization.

File: pycalphad/io/database.py [hunk 1]
```diff
@@ -33,6 +33,34 @@ def _to_tuple(lst):
     return tuple(_to_tuple(i) if isinstance(i, list) else i for i in lst)
 
 
+class Specie(object):
```
Comment (richardotis) [inline pycalphad/io/database.py]: The internet seems to say that "species" is spelled and pronounced the same way for the singular and plural case (like "sheep" or "deer"). I think we should compromise on clarity for the sake of grammatical correctness here.

File: pycalphad/io/database.py [hunk 2]
```diff
@@ -33,6 +33,34 @@ def _to_tuple(lst):
...
+
+    def __eq__(self, other):
+        """Two species are the same if their names and constituents are the same."""
+        if isinstance(other, self.__class__):
+            return (self.name == other.name) and (self.constituents == other.constituents)
+        else:
+            return False
+
+    def __hash__(self):
+        return hash(self.name)
```
Comment (richardotis) [inline pycalphad/io/database.py]: This is not a good implementation of hash because it will break putting species into sets and frozensets when some species have the same name but different charges. Or is the name an unambiguous descriptor?

File: pycalphad/io/database.py [hunk 3]
```diff
@@ -33,6 +33,34 @@ def _to_tuple(lst):
...
+        Dictionary of {element: quantity} where the element is a string and the quantity a float.
+    charge : int
+        Integer charge. Can be positive or negative.
+    """
+    def __init__(self, name, constituents, charge):
+        self.name = name
+        self.constituents = constituents
+        self.charge = charge
+
+    def __eq__(self, other):
```
Comment (richardotis) [inline pycalphad/io/database.py]: See comment for `__hash__`

File: pycalphad/io/database.py (hunk 1 above)
Comment (bocklund) [reply pycalphad/io/database.py]: Well that's embarrassing... I'll fix it.

File: pycalphad/io/database.py (hunk 2 above)
Comment (bocklund) [reply pycalphad/io/database.py]: The TDB format differentiates species (for example Al+2 and Al+3) by their names, which are arbitrary but unique and unambiguous. In that case, the names are conventionally `AL+3` and `AL+2` (but again, are arbitrary). 

If I am understanding your meaning correctly, the tests should demonstrate this case for the species 
```
SPECIES O-2                         O1/-2!
SPECIES O1                          O!
```

which have the same chemical stoichometry but are differentiated by name.

Comment (richardotis) [reply pycalphad/io/database.py]: That answers my question, so it should be fine.
